src/app/shared/utils/utils.ts

```diff
@@ -49,4 +51,40 @@ export class Utils {
     .replace(/%5C/g, '_')  // Replace encoded backslash with underscore
     .replace(/%20/g, '_'); // Replace encoded space with underscore
   }
+
+  /**
+   * Returns an estimated connection tier based on downlink.
+   * If downlink can't be found, it falls back to small image load time.
+   * 
+   * @returns {Promise<ConnectionTier>}
+   */
+  public async getConnectionTier(): Promise<ConnectionTier> {
+    // Try to retrieve the connection speed directly
+    const conn = (navigator as any).connection || (navigator as any).mozConnection || (navigator as any).webkitConnection;
+
+    if (conn?.downlink) {
+      const speed = conn.downlink;
+      if (speed < 1) return 'slow';
+      else if (speed < 10) return 'medium';
+      else if (speed < 100) return 'fast';
+      else return 'turbo';
+    }
+
+    // Image load time fallback
+    return new Promise<ConnectionTier>((resolve) => {
+      const img = new Image();
+      const start = performance.now();
+
+      img.onload = () => {
+        const duration = performance.now() - start;
+        if (duration > 1500) return resolve('slow');
+        else if (duration > 800) return resolve('medium');
+        else if (duration > 400) return resolve('fast');
+        else return resolve('turbo');
+      };
+
+      img.onerror = () => resolve('medium'); // assume average
+      img.src = 'https://www.google.com/images/phd/px.gif?' + Date.now();
```

Comment:
Curious about URL, I've never heard of this. Where did you get it?

Reply:
You can see an example of it here:
https://stackoverflow.com/questions/5529718/how-can-i-detect-internet-speed-in-javascript

Basically it's an extremely small file, like 40 bytes or something, and it's a 1x1 transparent pixel. It's hosted by Google so it's extremely reliable (up-time) as well. 

Finally, keep in mind that this is just a backup if the primary method of speed test doesn't work.